### PR TITLE
feat: email and phone validations

### PR DESCRIPTION
**Pull request by hrwX**
> Please provide enough information so that others can review your pull request:

> Explain the **details** for making this change. What existing problem does the pull request solve?

> Screenshots/GIFs

### CODE REVIEW

**Comment by sahil28297** [comment]
@hrwX @DeveloperMetal  This is big change, what should we do about the existing data that's already there is these fields, because if the data is not according to these validations, it will keep on throwing error with no context (the right way to fill these fields) to the users, might break during migration as well and we won't know how to fix it.

File: frappe/utils/__init__.py [hunk 1]
```diff
@@ -62,6 +62,23 @@ def get_email_address(user=None):
 
 	return frappe.db.get_value("User", user, "email")
 
+def validate_phone_number(phone_number, throw=False):
+	"""Returns True if valid phone number."""
+	if not phone_number:
+		return False
+
+	phone_number = phone_number.strip()
+	match = re.match(r"([0-9\ \+\_\-\,\.\*\#\(\)]){1,20}$", phone_number)
```
Comment (DeveloperMetal) [inline frappe/utils/__init__.py:71]: Not convinced this is the best way to store a normalized phone number. You can still get a wild set of different digits and symbol combinations. That's without event thinking of extension numbers.

 What do you think of only storing digits and spaces(to group numbers)? then, leave the formatting to the frontend?
 It does mean that we should update the frontend controllers to handle phone number formatting(among other things)

File: frappe/utils/__init__.py [hunk 2]
```diff
@@ -62,6 +62,23 @@ def get_email_address(user=None):
...
+
+	if not match and throw:
+		# frappe.throw(frappe._("{0} is not a valid Phone Number").format(phone_number), frappe.InvalidPhoneNumberError)
+		frappe.msgprint(frappe._("""
+			{0} is not a valid Phone Number.
+			<br><br>
+			Allowed charater are <b>+</b>, <b>-</b>, <b>()</b>, <b>0-9</b>
```
Comment (DeveloperMetal) [inline frappe/utils/__init__.py:78]: As mentioned before, the formatting is a bit too loose. And only throwing an exception from the backend will annoy users. We need a frontend update as well to auto format this data as it is being entered. Just good UX to not annoy users.

File: frappe/utils/__init__.py [hunk 3]
```diff
@@ -110,8 +127,14 @@ def _check(e):
...
 			if throw:
-				frappe.throw(frappe._("{0} is not a valid Email Address").format(e),
-					frappe.InvalidEmailAddressError)
+				# frappe.throw(frappe._("{0} is not a valid Email Address").format(e),
+				# 	frappe.InvalidEmailAddressError)
+				frappe.msgprint(frappe._("""
+					{0} is not a valid Email Address.
+					<br><br>
+					Should be of the format
+					johndoe@domain.com / johndoe@example.domain.com
```
Comment (DeveloperMetal) [inline frappe/utils/__init__.py:136]: Same comments as phone number validation applies here. Needs a frontend update avoid annoying user during input.

**Review by DeveloperMetal — CHANGES_REQUESTED** [review]
I agree with what @sahil28297 mentioned. If we are doing this we need to make sure we can normalize existing data on existing installations and somehow deal with invalid data already existing on these databases. Otherwise you can easily block an update.

Why not take a less strict path first?

- Warn user of records not complying with normalized rules for say the next couple of months?
- Eventually make that rule a hard requirement. Give our user base time to correct their data and the tools to find bad data

**Pull request closed without merging**